### PR TITLE
Add shortest distance to WeatherHandler

### PR DESCRIPTION
**Pull request by hmerk**

### CODE REVIEW

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java [hunk 1]
```diff
@@ -102,9 +131,25 @@ private void updateSensorValues() {
                 updateStatus(ThingStatus.ONLINE);
 
                 String content = response.getContentAsString();
-                JsonObject jsonResponse = gson.fromJson(content, JsonObject.class);
-                logger.debug("Response: {}", jsonResponse.toString());
-                // JsonElement observations = jsonResponse.get("value").getAsJsonArray().get(0);
+                if (content != null) {
+
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]:
```suggestion
```

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java [hunk 2]
```diff
@@ -102,9 +131,25 @@ private void updateSensorValues() {
...
+                        logger.debug("Response: {}", jsonResponse.toString());
+                        JsonElement observations = jsonResponse.get("value").getAsJsonArray().get(0);
+                        String rs = observations.toString();
+                        JsonObject jsonResult = gson.fromJson(rs, JsonObject.class);
+                        if (jsonResult != null) {
+                            JsonElement result = jsonResult.get("Observations").getAsJsonArray().get(0);
+                            JsonObject valueObj = gson.fromJson(result.toString(), JsonObject.class);
+                            if (valueObj != null) {
+                                JsonElement value = valueObj.get("result");
+                                logger.debug("Value: {}", value);
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: So can't you read the temperature and humidity here now? Why all these changes if we in the end still use fake values?

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java [hunk 3]
```diff
@@ -137,4 +182,68 @@ public void dispose() {
...
+                logger.debug("Retreived Content {}", content);
+                return content;
+            }
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+        }
+        return "";
+    }
+
+    private void getNearestOnlineLocation() {
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: I expect a return value from a get-method. Maybe rather rename it to `updateNearestOnlineLocation`?

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java [hunk 4]
```diff
@@ -137,4 +182,68 @@ public void dispose() {
...
+                            String position = coordinates.get(1).toString() + "," + coordinates.get(0).toString();
+                            Double distance = distanceCalculation(position);
+                            onlineLocations.put(id, distance);
+                        }
+                    }
+                }
+                Map<String, Double> sortedMap = onlineLocations.entrySet().stream().sorted(Map.Entry.comparingByValue())
+                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1,
+                                LinkedHashMap::new));
+                logger.debug("Nächste Messtelle @iot.id:{} / {}", sortedMap.keySet().iterator().next(),
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Please log in english.

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java [hunk 5]
```diff
@@ -137,4 +182,68 @@ public void dispose() {
...
+                        locationNames.get(sortedMap.keySet().iterator().next()));
+                logger.debug("Entfernung: {}", onlineLocations.get(sortedMap.keySet().iterator().next()));
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: dito

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java [hunk 6]
```diff
@@ -137,4 +182,68 @@ public void dispose() {
...
+            }
+        }
+    }
+
+    private Double distanceCalculation(String position) {
+
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]:
```suggestion
```

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java [hunk 7]
```diff
@@ -137,4 +182,68 @@ public void dispose() {
...
+                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1,
+                                LinkedHashMap::new));
+                logger.debug("Nächste Messtelle @iot.id:{} / {}", sortedMap.keySet().iterator().next(),
+                        locationNames.get(sortedMap.keySet().iterator().next()));
+                logger.debug("Entfernung: {}", onlineLocations.get(sortedMap.keySet().iterator().next()));
+            }
+        }
+    }
+
+    private Double distanceCalculation(String position) {
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Unusual method name - maybe better `getDistance`?

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java [hunk 8]
```diff
@@ -74,19 +88,34 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+                        bridgeHandler.config.refreshInterval, TimeUnit.SECONDS);
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
         }
     }
 
     private void updateSensorValues() {
+
+        getNearestOnlineLocation();
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Looks expensive to do this on every single refresh. Isn't it enough to get the closest station at initialization and then only update the list in case it goes offline?

**Review by kaikreuzer — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 2 above)
Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Comes with next commit

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 3 above)
Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Changed in next commit

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 4 above)
Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Changed in next commit

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 5 above)
Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Changed in next commit

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 8 above)
Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: No, what happens if they take the Thing offline and we don't get new values for a longer period?
I deeply thought about that, but there is no other option.

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 7 above)
Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Renamed with next commit

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 6 above)
Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Removed

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 1 above)
Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Removed

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 8 above)
Comment (Confectrician) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: How about only checking the online status of the sensor (via api call) at start of the update method and only get the nearest location in case it isn't available anymore?

As you said "they" will put it offline or in maintenance, so we will definetely gat some sort of informatoin from the api in any case.

Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Did not think about it that way.
Seems to be feasable.
@Confectrician Can you create the call to check if a given Thing is Online? Don't think we can go down to sensor level....

Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: @kaikreuzer @Confectrician 
Implemented 😉

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java [hunk 9]
```diff
@@ -74,67 +90,172 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
-                logger.debug("HTTP request failed with response code {}: {}", response.getStatus(),
-                        response.getReason());
+                logger.debug("Requesting URL {}", url);
+                logger.debug("Retreived Content {}", content);
+                return content;
             }
         } catch (InterruptedException | ExecutionException | TimeoutException e) {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
         }
+        return "";
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Wouldn't it better to return `null` here?

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java [hunk 10]
```diff
@@ -74,67 +90,172 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
-                logger.debug("Response: {}", jsonResponse.toString());
-                // JsonElement observations = jsonResponse.get("value").getAsJsonArray().get(0);
-                Double temperature = 11.3;
-                Double humidity = 0.85;
-
-                updateState(OpenSmartCityBindingConstants.CHANNEL_TEMPERATURE,
-                        new QuantityType<Temperature>(temperature, SIUnits.CELSIUS));
-                updateState(OpenSmartCityBindingConstants.CHANNEL_HUMIDITY, new DecimalType(humidity));
-
-            } else {
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java:115]: Why did you remove the else-clause?

**Review by kaikreuzer — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 9 above)
Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java]: Changed

File: bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java (hunk 10 above)
Comment (hmerk) [reply bundles/org.openhab.binding.opensmartcity/src/main/java/org/openhab/binding/opensmartcity/internal/OpenSmartCityWeatherHandler.java:115]: Got lost while moving stuff to apiRequest(String queryPath). Reimplemented with next commit

**Review by kaikreuzer — APPROVED** [review]
Thank you!
